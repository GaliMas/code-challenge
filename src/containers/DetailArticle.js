import { connect } from 'react-redux';
import { createArticle, fetchArticle, updateArticle, removeArticle } from '../actions/article';
import DetailArticle from '../views/DetailArticle';

const mapStateToProps = (state, ownProps) => ({
  id: ownProps.match.params.id,
  article: state.articles.articleDetailsById[ownProps.match.params.id],
});

const mapDispatchToProps = {
  createArticle,
  fetchArticle,
  updateArticle,
  removeArticle,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(DetailArticle);
