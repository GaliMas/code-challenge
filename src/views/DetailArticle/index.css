.detailArticle{
  padding: 100px 20px;
}
.detailArticle .labelDetailArticle {
  font-weight: bold;
}
.detailArticle .detailButtons{
  margin-top: 20px;
}
.detailArticle .detailRow{
  margin: 10px 0;
}

.detailArticle .btnRemove{
  margin: 10px;
}