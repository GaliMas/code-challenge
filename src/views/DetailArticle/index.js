import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Tag from '../../components/Tag';
import CreateArticleForm from '../CreateArticleForm';
import RemoveArticleForm from '../RemoveArticleForm';
import './index.css';

class DetailArticle extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isModalOpen: false,
      isEditMode: false,
      isRemoveMode: false,
    };
    this.onSave = this.onSave.bind(this);
    this.removeArticle = this.removeArticle.bind(this);
    this.toogleRemoveModal = this.toogleRemoveModal.bind(this);
    this.toogleEditModal = this.toogleEditModal.bind(this);
  }

  componentDidMount() {
    this.props.fetchArticle(this.props.id);
  }

  onSave(values) {
    if (this.state.isEditMode) {
      this.props.updateArticle(values);
    } else {
      this.props.createArticle(values);
    }
    this.toogleEditModal();
  }

  toogleEditModal() {
    this.setState(prevState => ({ isEditMode: !prevState.isEditMode }));
  }

  toogleRemoveModal() {
    this.setState(prevState => ({ isRemoveMode: !prevState.isRemoveMode }));
  }

  removeArticle() {
    this.props.removeArticle(this.props.article.id);
    this.toogleRemoveModal();
  }

  render() {
    const { id, author, title, content, published, tags } = this.props.article;
    return (
      <div className="detailArticle">
        <div className="detailRow">
          <span className="labelDetailArticle">Author: </span>
          <span className="contentDetail">{author}</span>
        </div>
        <div className="detailRow">
          <span className="labelDetailArticle">Title: </span>
          <span className="contentDetail">{title}</span>
        </div>
        <div className="detailRow">
          <span className="labelDetailArticle">Content: </span>
          <span className="contentDetail">{content}</span>
        </div>
        <span className="labelDetailArticle">{published ? 'Published' : 'Not published'}</span>
        <div className="detailRow">
          <span className="labelDetailArticle">Tags:</span>
          {tags && tags.map(tag => <Tag key={tag}>{tag}</Tag>)}
        </div>
        {this.state.isEditMode &&
          <CreateArticleForm
            onClose={this.toogleEditModal}
            onSave={this.onSave}
            id={id}
            author={author}
            title={title}
            content={content}
            published={published}
            tags={tags}
          />}
        {this.state.isRemoveMode
          && <RemoveArticleForm
            onClose={this.toogleRemoveModal}
            onRemove={this.removeArticle}
          />
        }
        <div className="detailButtons">
          <button
            className="btn btnEdit"
            onClick={this.toogleEditModal}
          >edit</button>
          <button
            className="btn btnRemove"
            onClick={this.toogleRemoveModal}
          >remove</button>
        </div>
      </div >
    );
  }
}

DetailArticle.propTypes = {
  article: PropTypes.shape({
    author: PropTypes.string,
    content: PropTypes.string,
    id: PropTypes.string,
    published: PropTypes.bool,
    tags: PropTypes.arrayOf(PropTypes.string),
    title: PropTypes.string,
  }),
  createArticle: PropTypes.func.isRequired,
  fetchArticle: PropTypes.func.isRequired,
  id: PropTypes.string,
  removeArticle: PropTypes.func.isRequired,
  updateArticle: PropTypes.func.isRequired,
};

DetailArticle.defaultProps = {
  article: {},
};

export default DetailArticle;
