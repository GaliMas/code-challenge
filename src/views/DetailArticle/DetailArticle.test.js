import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import DetailArticle from '.';

it('renders correctly', () => {
  const createArticle = jest.fn();
  const fetchArticle = jest.fn();
  const updateArticle = jest.fn();
  const removeArticle = jest.fn();
  const article = {
    author: 'Robb Farrell',
    content: 'ex molestias et. Harum placeat soluta officiis fuga.↵',
    id: '5467',
    published: false,
    tags: ['convergence', 'Run plum exploit'],
    title: 'Corporate Identity Director',
  };

  const tree = renderer.create(
    <DetailArticle
      article={article}
      createArticle={createArticle}
      fetchArticle={fetchArticle}
      updateArticle={updateArticle}
      removeArticle={removeArticle}
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should trigger onClose', () => {
  const createArticle = jest.fn();
  const fetchArticle = jest.fn();
  const updateArticle = jest.fn();
  const removeArticle = jest.fn();
  const article = {
    author: 'Robb Farrell',
    content: 'ex molestias et. Harum placeat soluta officiis fuga.↵',
    id: '5467',
    published: false,
    tags: ['convergence', 'Run plum exploit'],
    title: 'Corporate Identity Director',
  };
  const wrapper = mount(<DetailArticle
    createArticle={createArticle}
    fetchArticle={fetchArticle}
    updateArticle={updateArticle}
    removeArticle={removeArticle}
    article={article}
  />);
});
