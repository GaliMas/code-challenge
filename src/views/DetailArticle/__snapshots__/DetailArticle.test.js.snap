exports[`test renders correctly 1`] = `
<div
  className="detailArticle">
  <div
    className="detailRow">
    <span
      className="labelDetailArticle">
      Author: 
    </span>
    <span
      className="contentDetail">
      Robb Farrell
    </span>
  </div>
  <div
    className="detailRow">
    <span
      className="labelDetailArticle">
      Title: 
    </span>
    <span
      className="contentDetail">
      Corporate Identity Director
    </span>
  </div>
  <div
    className="detailRow">
    <span
      className="labelDetailArticle">
      Content: 
    </span>
    <span
      className="contentDetail">
      ex molestias et. Harum placeat soluta officiis fuga.â†µ
    </span>
  </div>
  <span
    className="labelDetailArticle">
    Not published
  </span>
  <div
    className="detailRow">
    <span
      className="labelDetailArticle">
      Tags:
    </span>
    <span
      className="tag">
      <span
        className="tagText">
        convergence
      </span>
    </span>
    <span
      className="tag">
      <span
        className="tagText">
        Run plum exploit
      </span>
    </span>
  </div>
  <div
    className="detailButtons">
    <button
      className="btn btnEdit"
      onClick={[Function]}>
      edit
    </button>
    <button
      className="btn btnRemove"
      onClick={[Function]}>
      remove
    </button>
  </div>
</div>
`;
